PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/ui/SparkSubmissionJobUploadStorageGen2Card.kt

```diff
@@ -75,18 +88,59 @@ class SparkSubmissionJobUploadStorageGen2Card : SparkSubmissionJobUploadStorageB
             }
         }
 
-        layout = formBuilder.createGridLayoutManager()
-        formBuilder.allComponentConstraints.forEach { (component, gridConstrains) -> add(component, gridConstrains) }
+        formBuilder.buildPanel()
     }
 
-    override val title = SparkSubmitStorageType.ADLS_GEN2.description
+    inner class ViewModel: SparkSubmissionJobUploadStorageBasicCard.ViewModel() {
+        val rootUri: AbfsUri?
+            get() {
+                val rootPathText = gen2RootPathField.text?.trim()
+                return if (AbfsUri.isType(rootPathText)) AbfsUri.parse(rootPathText) else null
+            }
+
+        override var cluster: IClusterDetail?
+            get() = super.cluster
+            set(cluster) {
+                val gen2Account = cluster?.storageAccount?.name ?: return
+
+                if (!storageKeyField.text.isNullOrBlank()) {
```

Comment:
Why do we need to set storage key field when we set cluster? I suppose this behaviour should be done in `setData` method.

Reply:
Here the setData/getData have a role changes in MVVM. They are used for exchanging between Model and ViewModel ( Model <=> ViewModel). The ViewModel can operate view directly.